qadence/constructors/ansatze.py

```diff
@@ -320,3 +320,222 @@ def hea_bDAQC(*args: Any, **kwargs: Any) -> Any:
+        support = tuple(range(n_qubits))
+
+    alt_func_dict = {
+        Strategy.DIGITAL: alt_digital,
+        Strategy.SDAQC: alt_sDAQC,
+        Strategy.BDAQC: alt_bDAQC,
+        Strategy.ANALOG: alt_analog,
+    }
+
+    try:
+        alt_func = alt_func_dict[strategy]
+    except KeyError:
+        raise KeyError(f"Strategy {strategy} not recognized.")
+
+    hea_block: AbstractBlock = alt_func(
+        n_qubits=n_qubits,
+        m_block_qubits=m_block_qubits,
+        depth=depth,
+        param_prefix=param_prefix,
+        support=support,
+        **strategy_args,
+    )  # type: ignore
+
+    return hea_block
+
+
+#################
+## DIGITAL ALT ##
+#################
+
+
+def _entanglers_block_digital(
+    n_qubits: int,
+    m_block_qubits: int,
+    depth: int,
+    param_prefix: str = "theta",
+    support: tuple[int, ...] = None,
+    entangler: Type[DigitalEntanglers] = CNOT,
+) -> list[AbstractBlock]:
+    if support is None:
+        support = tuple(range(n_qubits))
+    iterator = itertools.count()
+    ent_list: list[AbstractBlock] = []
+
+    for d in range(depth):
+        ents = []
+        if not d % 2:
+            ents.append(
+                kron(
+                    _entangler(
+                        control=support[i + j],
+                        target=support[i + j + 1],
+                        param_str=param_prefix + f"_ent_{next(iterator)}",
+                        op=entangler,
+                    )
+                    for i in range(0, n_qubits, m_block_qubits)
+                    for j in range(0, m_block_qubits, 2)
+                    if i + j + 1 < n_qubits and j + 1 < m_block_qubits
+                )
+            )
+            if m_block_qubits > 2:
+                ents.append(
+                    kron(
+                        _entangler(
+                            control=support[i + j],
+                            target=support[i + j + 1],
+                            param_str=param_prefix + f"_ent_{next(iterator)}",
+                            op=entangler,
+                        )
+                        for i in range(0, n_qubits, m_block_qubits)
+                        for j in range(1, m_block_qubits, 2)
+                        if i + j + 1 < n_qubits and j + 1 < m_block_qubits
+                    )
+                )
+
+        elif d % 2:
+            ents.append(
+                kron(
+                    _entangler(
+                        control=support[i + j],
+                        target=support[i + j + 1],
+                        param_str=param_prefix + f"_ent_{next(iterator)}",
+                        op=entangler,
+                    )
+                    for i in range(-m_block_qubits // 2, n_qubits, m_block_qubits)
+                    for j in range(0, m_block_qubits, 2)
+                    if i + j + 1 < n_qubits and j + 1 < m_block_qubits and i + j >= 0
+                )
+            )
+            if m_block_qubits > 2:
```

Comment:
Given the repeating blocks of code, I would suggest a syntax change by introducing at the beginning of the loop:
- `start_i, start_j` variables to be set depending on the `d %2` case and `m_block_qubits`
- entangler_indices = [(i,j) for i in range(start_i, n_qubits, m_block_qubits) for j in range(start_j, m_block_qubits, 2) if ...] and adjust the if conditions accordingly
- then you will only write the `kron(_entangler(...) for i,j in entangler_indices)` once!

Reply:
see https://github.com/pasqal-io/qadence/pull/565#discussion_r1767045762